File: TableRender/Converter.php
<?php
error_reporting(0);
class Paper
{

    public $tables;
    public $pmid;
    private $file_json;
    public $annotations = [];
    public $relations = [];

    function __construct($file_input)
    {
        $this->file_json = file_get_contents($file_input);
        $this->file_json = json_decode($this->file_json, true, 512, JSON_UNESCAPED_UNICODE);
        $this->pmid = $_GET['pmcid'];
        $this->tables = [];
        foreach ($this->file_json["documents"] as $table_json) {
            $new_table = new Table($table_json);
            array_push($this->tables, $new_table);
            $this->collect_table_annotations($new_table);
            $this->relations = array_merge($this->relations, $new_table->table_relations);
        }
    }

    private function collect_table_annotations($table)
    {
        if ($table->title_annotations)
            $this->annotations = array_merge($this->annotations, $table->title_annotations);
        if ($table->caption_annotations)
            $this->annotations = array_merge($this->annotations, $table->caption_annotations);
        if ($table->table_annotations)
            $this->annotations = array_merge($this->annotations, $table->table_annotations);
        if ($table->footer_annotations)
            $this->annotations = array_merge($this->annotations, $table->footer_annotations);
    }

    public function outputTables()
    {
        $return_str = "<a href=https://www.ncbi.nlm.nih.gov/pmc/articles/{$this->pmid} target=_newWindow>{$this->pmid}</a>";
        foreach ($this->tables as $table) {
            $return_str .= $table->output_table();
        }
        return $return_str;
    }
}

class Table
{

    public $table_num = 0;
    public $headers;
    public $data_sections = [];
    public $caption;
    public $caption_annotations;
    public $title;
    public $title_annotations = [];
    public $table_annotations = [];
    public $table_relations = [];
    public $annotation_ids = [];
    public $footer;
    public $footer_annotations = [];
    public $max_col_size = 1;
    private $caption_relations;
    private $caption_annots;
    private $title_relations;
    private $footer_relations;

    private $piped_indexes = [];

    function __construct($json_data)
    {
        $this->table_num = $json_data["id"];
        $this->data_sections = $this->__get_data_sections($json_data);
        $this->headers = $this->__expand_headers($json_data["passages"][2]["column_headings"]);
        $this->caption = $json_data["passages"][1]["text"];
        $this->caption_annots = $json_data["passages"][1]["annotations"]; // this line is problematic?
        $this->caption_relations = $json_data["passages"][1]["relations"];
        $this->title = $json_data["passages"][0]["text"];
        $this->title_annotations = $json_data["passages"][0]["annotations"];
        $this->title_relations = $json_data["passages"][1]["relations"];
        $this->footer = $json_data["passages"][3]["text"];
        $this->footer_annotations = $json_data["passages"][3]["annotations"];
        $this->footer_relations = $json_data["passages"][1]["relations"];
    }

    private function __expand_headers($headers)
    {
        // Expand piped columns to appear more like the original article
        $new_headers = []; // 2D array for rows and columns
        $used_strings = [];
        $max_row_count = sizeof($headers);
        for ($i = 0; $i < sizeof($headers); $i++) { // $i == cell
            $used_strings[] = [];
            $header = $headers[$i];//current header
            if (strpos($header["cell_text"], "|")) {
                $pipe_split = explode("|", $header["cell_text"]); // split by pipe character
                $this->piped_indexes[] = $i;
                if (sizeof($pipe_split) > $max_row_count)
                    $max_row_count = sizeof($pipe_split);
                for ($o = 0; $o < sizeof($pipe_split); $o++) { // iterate over each string
                    $sub_string = $pipe_split[$o];
                    $new_header = $header;
                    if (in_array($sub_string, $used_strings[$i])) // ignore duplicates after first occurrence
                        $new_header["cell_text"] = "";
                    else {
                        $used_strings[$i][] = $sub_string; // log the used string
                        // create new header record with new substring
                        $new_header["cell_text"] = $sub_string;
                    }
                    $new_headers[$o][$i] = $new_header; // add the new header row/cell
                }
            } else {
                $new_headers[0][$i] = $header; // no split needed, add the current header cell to the top row
            }
        }
        // fill any gaps in rows with blank cells
        for ($i = 0; $i < sizeof($new_headers); $i++) {
            if (sizeof($new_headers[$i]) < sizeof($headers)) {
                $blank_cell = $new_headers[$i][0];
                $blank_cell["cell_text"] = "";
                for ($o = 0; $o < sizeof($headers); $o++) {
                    if (!array_key_exists($o, $new_headers[$i])) {
                        $new_headers[$i][$o] = $blank_cell;
                    }
                }
            }
        }
        return $new_headers;
    }

    private function __get_data_sections($json_data)
    {
        $data_sections = [];
        foreach ($json_data["passages"][2]["data_section"] as $section) {
            $rows = [];
            foreach ($section["data_rows"] as $row) {
                if (sizeof($row) > $this->max_col_size)
                    $this->max_col_size = sizeof($row);
                $new_row = [];
                foreach ($row as $cell) {
                    $new_row[] = [$cell["cell_id"], $cell["cell_text"]];
                }
                $rows[] = $new_row;
            }
            $data_sections[] = [$section["text"], $rows];
        }
        foreach ($json_data["passages"][2]["annotations"] as $annotation) {
            $this->table_annotations[] = $annotation;
            $this->annotation_ids[] = $annotation["locations"][0]["cell_id"];
        }
        foreach ($json_data["passages"][2]["relations"] as $relation) {
            $this->table_relations[] = $relation;
        }
        return $data_sections;
    }

    private function __get_headers_output()
    {
        $return_str = "";
        foreach ($this->headers as $row) {
            $return_str .= "<tr>";
            for ($i = 0; $i < sizeof($row); $i++) {
                $cell = $row[$i];
                if (in_array($cell["cell_id"], $this->annotation_ids)) {
                    $cell_annot_id = "";
                    foreach ($this->table_annotations as $annot) {
                        if ($annot["locations"][0]["cell_id"] == $cell["cell_id"]) {
                            $cell_annot_id = $annot["id"];
                            break;
                        }
                    }
                    $cell_content = "<div class='annotation' data-annot-id='{$cell_annot_id}'>{$cell["cell_text"]}</div>";
                } else {
                    $cell_content = "<div data-annot-id='{$cell["cell_id"]}'>{$cell["cell_text"]}</div>";
                }
                // calculate col span.
                $col_span = 1;
                // look ahead at remaining columns in row
                for ($o = $i + 1; $o < sizeof($row); $o++) {
                    if (($cell["cell_text"] == $row[$o]["cell_text"]) || (!in_array($o, $this->piped_indexes)
                            && !in_array($i, $this->piped_indexes)
                            && ($cell["cell_text"] == "" || $row[$o]["cell_text"] == ""))) {
                        $col_span++;
                    } else {
                        break;
                    }
                }
                $return_str .= "<th colspan='{$col_span}'>
                        $cell_content
                    </th>";
                // skip detected duplicate headers
                $i += $col_span - 1;
            }
            $return_str .= "</tr>";
        }
        return $return_str;
    }

    private function __get_caption_output()
    {
        if ($this->caption_annotations) {
            $added_char_length = 0;
            foreach ($this->caption_annotations as $annot) {
                $replace_str = "<div class='annotation' data-annot-id='{$annot['id']}'>{$annot['text']}</div>";
                $this->caption = substr_replace($this->caption, $replace_str, $annot["locations"][0]["offset"] + $added_char_length, $annot["locations"][0]["length"]);
                $added_char_length += (strlen($replace_str) - $annot["locations"][0]["length"]);
            }
        }
        return "<caption style='text-align:left'><br/><b>{$this->title}</b><p>{$this->caption}</p></caption>";
    }

    private function __get_table_num_output()
    {
        return "
            <p>Auto-CORPus id: {$this->table_num}</p>
        ";
    }

    private function __get_sections_output()
    {
        $return_str = "";
        foreach ($this->data_sections as $section) {
            if ($section[0]) {
                $header_count = $this->max_col_size;
                $return_str .= "<tr><td colspan='{$header_count}'><b>{$section[0]}<b></td></tr>";
            }
            foreach ($section as $new_section) {
                foreach($new_section as $row) {
                    $return_str .= "<tr>";
                    for($i = 0; $i < sizeof($row); $i++) {
                        $cell = $row[$i];
                        if (in_array($cell[0], $this->annotation_ids)) {
                            $cell_annot_id = "";
                            foreach ($this->table_annotations as $annot) {
                                if ($annot["locations"][0]["cell_id"] == $cell[0]) {
                                    $cell_annot_id = $annot["id"];
                                    break;
                                }
                            }
                            $cell_content = "<div class='annotation' data-annot-id='{$cell_annot_id}'>{$cell[1]}</div>";
                        } else {
                            $cell_content = "<div data-annot-id='{$cell[0]}'>{$cell[1]}</div>";
                        }
                        $return_str .= "<td>
                            $cell_content
                        </td>";
                    }
                    $return_str .= "</tr>";
                }
            }
        }
        return $return_str;
    }

    public function output_table()
    {
        $return_str = $this->__get_table_num_output();
        $return_str .= $this->__get_caption_output();
        $return_str .= "<div class='table-container'><table border=1 width='80%'><thead>";
        $return_str .= $this->__get_headers_output();
        $return_str .= "</thead><tbody>";
        $return_str .= $this->__get_sections_output();
        $return_str .= "</tbody></table></div>";
        if ($this->table_footer) {
            $return_str .= "<p>{$this->table_footer}</p>";
        }
        return $return_str;
    }
}

function mb_substr_replace($original, $replacement, $position, $length)
{
    $startString = mb_substr($original, 0, $position, "UTF-8");
    $endString = mb_substr($original, $position + $length, mb_strlen($original), "UTF-8");

    $out = $startString . $replacement . $endString;

    return $out;
}

File: converter.py
import json
from urllib.parse import quote



class Paper:
    def __init__(self, file_input: str, pmcid: str):
        with open(file_input, encoding="utf-8") as f:
            self.file_json = json.load(f)
        self.pmid = pmcid
        self.tables = []
        self.annotations = []
        self.relations = []

        for table_json in self.file_json["documents"]:
            new_table = Table(table_json)
            self.tables.append(new_table)
            self._collect_table_annotations(new_table)
            self.relations.extend(new_table.table_relations)

    def _collect_table_annotations(self, table):
        if table.title_annotations:
            self.annotations.extend(table.title_annotations)
        if table.caption_annotations:
            self.annotations.extend(table.caption_annotations)
        if table.table_annotations:
            self.annotations.extend(table.table_annotations)
        if table.footer_annotations:
            self.annotations.extend(table.footer_annotations)

    def output_tables(self) -> str:
        link = f'<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/{quote(self.pmid)}" target="_blank">{self.pmid}</a>'
        return link + ''.join([table.output_table() for table in self.tables])

class Table:
    def __init__(self, json_data):
        self.table_num = json_data.get("id", "unknown")
        self.piped_indexes = []
        self.max_col_size = 1
        self.annotation_ids = []

        passages = json_data.get("passages", [])
        
        # Safe extraction helper
        def get_passage(idx):
            return passages[idx] if len(passages) > idx else {}

        title_passage = get_passage(0)
        caption_passage = get_passage(1)
        table_passage = get_passage(2)
        footer_passage = get_passage(3)

        # Title
        self.title = title_passage.get("text", "")
        self.title_annotations = title_passage.get("annotations", [])

        # Caption
        self.caption = caption_passage.get("text", "")
        self.caption_annotations = caption_passage.get("annotations", [])
        self.caption_relations = caption_passage.get("relations", [])

        # Footer
        self.footer = footer_passage.get("text", "")
        self.footer_annotations = footer_passage.get("annotations", [])
        self.footer_relations = footer_passage.get("relations", [])

        # Headers
        self.headers = self._expand_headers(table_passage.get("column_headings", []))
        self.table_annotations = table_passage.get("annotations", [])
        self.table_relations = table_passage.get("relations", [])
        self.annotation_ids = [a["locations"][0]["cell_id"] for a in self.table_annotations]

        # Data sections
        self.data_sections = self._get_data_sections(table_passage)



    def _expand_headers(self, headers):
        new_headers = []
        used_strings = []
        max_row_count = len(headers)
        for i, header in enumerate(headers):
            used_strings.append([])
            if '|' in header["cell_text"]:
                pipe_split = header["cell_text"].split('|')
                self.piped_indexes.append(i)
                max_row_count = max(max_row_count, len(pipe_split))
                for o, sub_string in enumerate(pipe_split):
                    new_header = header.copy()
                    if sub_string in used_strings[i]:
                        new_header["cell_text"] = ""
                    else:
                        used_strings[i].append(sub_string)
                        new_header["cell_text"] = sub_string
                    if len(new_headers) <= o:
                        new_headers.append({})
                    new_headers[o][i] = new_header
            else:
                if len(new_headers) == 0:
                    new_headers.append({})
                new_headers[0][i] = header

        for i, row in enumerate(new_headers):
            for o in range(len(headers)):
                if o not in row:
                    blank_cell = headers[o].copy()
                    blank_cell["cell_text"] = ""
                    row[o] = blank_cell
            new_headers[i] = [row[o] for o in sorted(row)]

        return new_headers

    def _get_data_sections(self, json_data):
        data_sections = []
        for section in json_data["data_section"]:
            print(section)
            rows = []
            for row in section["data_rows"]:
                self.max_col_size = max(self.max_col_size, len(row))
                new_row = [[cell["cell_id"], cell["cell_text"]] for cell in row]
                rows.append(new_row)
            data_sections.append(["", rows])

        return data_sections

    def _get_headers_output(self):
        html = ""
        for row in self.headers:
            html += "<tr>"
            i = 0
            while i < len(row):
                cell = row[i]
                cell_id = cell.get("cell_id", "")
                content = cell["cell_text"]
                if cell_id in self.annotation_ids:
                    annot_id = next((a["id"] for a in self.table_annotations if a["locations"][0]["cell_id"] == cell_id), "")
                    cell_content = f"<div class='annotation' data-annot-id='{annot_id}'>{content}</div>"
                else:
                    cell_content = f"<div data-annot-id='{cell_id}'>{content}</div>"
                col_span = 1
                for o in range(i + 1, len(row)):
                    next_cell = row[o]
                    if content == next_cell["cell_text"] or (i not in self.piped_indexes and o not in self.piped_indexes and (not content or not next_cell["cell_text"])):
                        col_span += 1
                    else:
                        break
                html += f"<th colspan='{col_span}'>{cell_content}</th>"
                i += col_span
            html += "</tr>"
        return html

    def _get_caption_output(self):
        caption_text = self.caption
        added_char_length = 0
        for annot in self.caption_annotations:
            offset = annot["locations"][0]["offset"] + added_char_length
            length = annot["locations"][0]["length"]
            replace_str = f"<div class='annotation' data-annot-id='{annot['id']}'>{annot['text']}</div>"
            caption_text = caption_text[:offset] + replace_str + caption_text[offset + length:]
            added_char_length += len(replace_str) - length
        return f"<caption style='text-align:left'><br/><b>{self.title}</b><p>{caption_text}</p></caption>"

    def _get_table_num_output(self):
        return f"<p>Auto-CORPus id: {self.table_num}</p>"

    def _get_sections_output(self):
        html = ""
        for section in self.data_sections:
            section_title, rows = section
            if section_title:
                html += f"<tr><td colspan='{self.max_col_size}'><b>{section_title}</b></td></tr>"
            for row in rows:
                html += "<tr>"
                for cell_id, cell_text in row:
                    if cell_id in self.annotation_ids:
                        annot_id = next((a["id"] for a in self.table_annotations if a["locations"][0]["cell_id"] == cell_id), "")
                        cell_content = f"<div class='annotation' data-annot-id='{annot_id}'>{cell_text}</div>"
                    else:
                        cell_content = f"<div data-annot-id='{cell_id}'>{cell_text}</div>"
                    html += f"<td>{cell_content}</td>"
                html += "</tr>"
        return html

    def output_table(self):
        html = self._get_table_num_output()
        html += self._get_caption_output()
        html += "<div class='table-container'><table border=1 width='80%'><thead>"
        html += self._get_headers_output()
        html += "</thead><tbody>"
        html += self._get_sections_output()
        html += "</tbody></table></div>"
        if hasattr(self, 'table_footer') and self.table_footer:
            html += f"<p>{self.table_footer}</p>"
        return html
